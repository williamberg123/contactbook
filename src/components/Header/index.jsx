import { useContext } from 'react';
import { Link } from 'react-router-dom';
import AppContext from '../../contexts/AppProvider/AppContext';
import RenderIf from '../RenderIf';
import StyledHeader from './styles';

export default function Header() {
	const { user } = useContext(AppContext);

	return (
		<StyledHeader>
			<span>AGENDA</span>
			<div className="links-div">
				<RenderIf isTrue={ !user }>
						<Link to="/login">Entrar</Link>
						<Link to="/createaccount">Cadastrar</Link>
				</RenderIf>
				<RenderIf isTrue={ !!user }>
					<Link to="/createcontact">Criar contato</Link>
					<Link to="/logout">Sair</Link>
				</RenderIf>
			</div>
		</StyledHeader>
	);
}
