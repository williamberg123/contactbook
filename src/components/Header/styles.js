import { useContext } from 'react';
import styled, { css } from 'styled-components';
import AppContext from '../../contexts/AppProvider/AppContext';

const StyledHeader = styled.div`
	width: 100%;
	display: flex;
	align-items: center;
	justify-content: space-between;
	padding: 30px 20px;
	background-color: #0084d6;
	color: white;
	font-weight: bold;

	${() => {
		const { user } = useContext(AppContext);

		return css`
			padding: ${ user ? '20px' : '30px 20px' };
		`;
	}}

	& .links-div {
		display: flex;
		align-items: center;
	}

	& .links-div .user-profile {
		position: relative;
		font-weight: normal;
		border-radius: 50%;
		background-color: #005b93;
		color: white;
		display: flex;
		align-items: center;
		justify-content: center;
		margin-left: 10px;
		cursor: pointer;
		transition: 0.4s;
	}

	& .links-div .user-profile .hidden-menu {
		position: absolute;
		font-size: 0.7rem;
		list-style: none;
		display: none;
		flex-direction: column;
		align-items: center;
		bottom: -75px;
		right: -50%;
		padding: 5px;
		background-color: #005b93;
		border-radius: 5px;
	}

	& .links-div .user-profile:hover .hidden-menu {
		display: flex;
	}

	& .links-div .user-profile:hover .hidden-menu .triangle {
		position: absolute;
		top: -28px;
		right: 20px;
		width: 0;
		height: 0;
		border-color: transparent;
		border-width: 10px;
		border-style: solid;
		border-bottom-color: #005b93;
	}

	& .links-div .user-profile .hidden-menu div {
		width: 70px;
		text-align: center;
		margin: 5px 0;
		padding: 2px;
	}

	& .links-div .user-profile .hidden-menu .account-actions:hover {
		background-color: #0084d6;
	}

	& .links-div .user-first-letter {
		width: 35px;
		height: 35px;
		display: flex;
		align-items: center;
		justify-content: center;
		border-radius: 50%;
	}

	& .links-div .user-first-letter:hover {
		background-color: #0074b2;
		text-decoration: none;
	}

	& a {
		text-decoration: none;
		font-weight: lighter;
		color: white;
		margin: 0 10px;
		transition: 0.2s;
	}

	& a:hover {
		color: #75d1ff;
		text-decoration: underline;
	}
`;

export default StyledHeader;
