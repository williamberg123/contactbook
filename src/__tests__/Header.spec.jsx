import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { describe, expect, it } from 'vitest';
import Header from '../components/Header';
import AppContext from '../contexts/AppProvider/AppContext';

describe('<Header />', () => {
	it('should render span with the text "AGENDA"', () => {
		render(
			<BrowserRouter>
				<AppContext.Provider value={{ user: null }}>
						<Header />
				</AppContext.Provider>,
			</BrowserRouter>,
		);

		expect(screen.getByText(/agenda/i)).toBeInTheDocument();
	});

	it('should render signin and register links when there is no loggedin user', () => {
		expect.assertions(6);

		render(
			<BrowserRouter>
				<AppContext.Provider value={{ user: null }}>
						<Header />
				</AppContext.Provider>,
			</BrowserRouter>,
		);

		expect(screen.getByRole('link', { name: /entrar/i })).toBeInTheDocument();
		expect(screen.getByRole('link', { name: /cadastrar/i })).toBeInTheDocument();

		expect(screen.getByRole('link', { name: /entrar/i })).toHaveAttribute('href', '/login');
		expect(screen.getByRole('link', { name: /cadastrar/i })).toHaveAttribute('href', '/createaccount');

		expect(screen.queryByRole('link', { name: /criar contato/i })).not.toBeInTheDocument();
		expect(screen.queryByRole('img')).not.toBeInTheDocument();
	});

	it('should render criar contato link and profile image when there is a loggedin user', () => {
		expect.assertions(3);

		const context = {
			user: {
				email: 'williamberg567@gmail.com',
				photoURL: 'https://lh3.googleusercontent.com/a-/AFdZucreO7tbhF4l_S5dyGaniu68GR3ftpjaE3RYVvUVeQ=s96-c',
			},
		};

		render(
			<BrowserRouter>
				<AppContext.Provider value={context}>
						<Header />
				</AppContext.Provider>,
			</BrowserRouter>,
		);

		expect(screen.getByRole('link', { name: /criar contato/i })).toBeInTheDocument();
		expect(screen.getByRole('link', { name: /criar contato/i })).toHaveAttribute('href', '/newcontact');
		expect(screen.getByRole('img')).toHaveAttribute('src', context.photoURL);
	});

	it('should render hidden options when profile is hover', () => {
		const context = {
			user: {
				email: 'williamberg567@gmail.com',
				photoURL: 'https://lh3.googleusercontent.com/a-/AFdZucreO7tbhF4l_S5dyGaniu68GR3ftpjaE3RYVvUVeQ=s96-c',
			},
		};

		render(
			<BrowserRouter>
				<AppContext.Provider value={context}>
						<Header />
				</AppContext.Provider>,
			</BrowserRouter>,
		);

		const userProfile = document.querySelector('.user-profile');

		expect(userProfile).toBeInTheDocument();
	});

	it('should match snapshot', () => {
		render(
			<BrowserRouter>
				<AppContext.Provider value={{ user: null }}>
						<Header />
				</AppContext.Provider>,
			</BrowserRouter>,
		);

		expect(screen.getByText(/agenda/i).parentElement).toMatchSnapshot();
	});
});
