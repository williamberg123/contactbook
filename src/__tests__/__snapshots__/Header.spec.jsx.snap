// Vitest Snapshot v1

exports[`<Header /> > should match snapshot 1`] = `
.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 30px 20px;
  background-color: #0084d6;
  color: white;
  font-weight: bold;
  padding: 30px 20px;
}

.c0 img {
  width: 40px;
  height: 40px;
  border-radius: 50%;
}

.c0 .links-div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .links-div .user-profile {
  position: relative;
  font-weight: normal;
  border-radius: 50%;
  background-color: #005b93;
  color: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-left: 10px;
  cursor: pointer;
  -webkit-transition: 0.4s;
  transition: 0.4s;
}

.c0 .links-div .user-profile .hidden-menu {
  position: absolute;
  font-size: 0.7rem;
  list-style: none;
  display: none;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  bottom: -75px;
  right: -50%;
  padding: 5px;
  background-color: #005b93;
  border-radius: 5px;
}

.c0 .links-div .user-profile:hover .hidden-menu {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 .links-div .user-profile:hover .hidden-menu .triangle {
  position: absolute;
  top: -28px;
  right: 20px;
  width: 0;
  height: 0;
  border-color: transparent;
  border-width: 10px;
  border-style: solid;
  border-bottom-color: #005b93;
}

.c0 .links-div .user-profile .hidden-menu div {
  width: 70px;
  text-align: center;
  margin: 5px 0;
  padding: 2px;
}

.c0 .links-div .user-profile .hidden-menu .account-actions:hover {
  background-color: #0084d6;
}

.c0 .links-div .user-first-letter {
  width: 35px;
  height: 35px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 50%;
}

.c0 .links-div .user-first-letter:hover {
  background-color: #0074b2;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 a {
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: lighter;
  color: white;
  margin: 0 10px;
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c0 a:hover {
  color: #75d1ff;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<header
  class="c0"
>
  <span>
    AGENDA
  </span>
  <div
    class="links-div"
  >
    <a
      href="/login"
    >
      Entrar
    </a>
    <a
      href="/createaccount"
    >
      Cadastrar
    </a>
  </div>
</header>
`;
